### ISSUE TITLE
Broken S3 Bucket Listing with Strict Access Policies

### ISSUE THREAD

**Issue by darthShadow** [rclone/rclone#4990]
#### What is the problem you are having with rclone?

S3 bucket listing is broken with 1.54 when strict access policies are used.

#### What is your rclone version (output from `rclone version`)

```
rclone v1.54.0
- os/arch: linux/amd64
- go version: go1.15.7
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Ubuntu 18.04.5

####  Which cloud storage system are you using? (e.g. Google Drive)

Minio S3

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

```
rclone-v1.54.0 --s3-no-check-bucket --config=/home/darthshadow/rclone/rclone.conf lsd ZenStorage-Inbound:zd-movies-non-english/movies-non-english/
```

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
2021/02/02 16:52:57 DEBUG : rclone: Version "v1.54.0" starting with parameters ["rclone-v1.54.0" "--s3-no-check-bucket" "--config=/home/darthshadow/rclone/rclone.conf" "lsd" "ZenStorage-Inbound:zd-movies-non-english/movies-non-english/" "-vv"]
2021/02/02 16:52:57 DEBUG : Using config file from "/home/darthshadow/rclone/rclone.conf"
2021/02/02 16:52:57 DEBUG : Creating backend with remote "ZenStorage-Inbound:zd-movies-non-english/movies-non-english/"
2021/02/02 16:52:57 Failed to create file system for "ZenStorage-Inbound:zd-movies-non-english/movies-non-english/": Forbidden: Forbidden
        status code: 403, request id: 165FFC768E1DA1F7, host id:
```

---

The access policies allow the user to only access a specific subpath in the bucket like: `zd-movies-non-english/movies-non-english/German/movies/`. There are multiple such policies for specific paths in the bucket.

This works perfectly with 1.53.4.

```
2021/02/02 17:15:31 DEBUG : rclone: Version "v1.53.4" starting with parameters ["rclone-v1.53.4" "--s3-no-check-bucket" "--config=/home/darthshadow/rclone/rclone.conf" "lsd" "ZenStorage-Inbound:zd-movies-non-english/movies-non-english/" "-vv"]
2021/02/02 17:15:31 DEBUG : Using config file from "/home/darthshadow/rclone/rclone.conf"
2021/02/02 17:15:31 DEBUG : Creating backend with remote "ZenStorage-Inbound:zd-movies-non-english/movies-non-english/"
2021/02/02 17:15:31 DEBUG : fs cache: renaming cache item "ZenStorage-Inbound:zd-movies-non-english/movies-non-english/" to be canonical "ZenStorage-Inbound:zd-movies-non-english/movies-non-english"
           0 2021-02-02 17:15:31        -1 Bollywood
           0 2021-02-02 17:15:31        -1 Danish
           0 2021-02-02 17:15:31        -1 Dutch
           0 2021-02-02 17:15:31        -1 German
           0 2021-02-02 17:15:31        -1 Movies-4k-Nordic
           0 2021-02-02 17:15:31        -1 Swedish
2021/02/02 17:15:31 DEBUG : 3 go routines active
```

Dumping the HTTP Requests shows an extra request from `1.53.4` which is missing in `1.54`:

```
GET /zd-movies-non-english?delimiter=%2F&max-keys=1000&prefix=movies-non-english%2F HTTP/1.1
```

**Comment by ncw** [comment]
> Dumping the HTTP Requests shows an extra request from `1.53.4` which is missing in `1.54`:
> 
> ```
> GET /zd-movies-non-english?delimiter=%2F&max-keys=1000&prefix=movies-non-english%2F HTTP/1.1
> ```

That is rclone listing the bucket with the prefix `movies-non-english`

As far as I can see 1.53 does an identical request?

Can you post `-vv --dump bodies` for 1.54 and 1.53.4 ?

**Comment by darthShadow** [comment]
> As far as I can see 1.53 does an identical request?

1.53 does the request and fetches the correct listing. 1.54 doesn't do that request and bails out with a "Forbidden" error.

[1_53_4.requests.txt](https://github.com/rclone/rclone/files/5913553/1_53_4.requests.txt)
[1_54_0.requests.txt](https://github.com/rclone/rclone/files/5913554/1_54_0.requests.txt)

**Comment by ncw** [comment]
OK I see what you mean now. Both 1.53.4 and 1.54.0 do `HEAD /zd-movies-non-english/movies-non-english`.

In both cases it returns with `403 Forbidden`

However 1.54.0 stops at that point and returns the error message to the user, wheras 1.53.4 carries on.

I suspect if rclone was to do `HEAD /zd-movies-non-english/movies-non-english/` it would succeed with your rules - is that correct?

Can you relax the rules so that this would work  `HEAD /zd-movies-non-english/movies-non-english`?

I'm not sure which commit introduced this problem - can you `git bisect` it for me?

**Comment by brvi** [comment]
I think that if you relax the rules, technically you can open up a vector for exploring data that doesn't belong to that user.

If @darthShadow relaxes their rules to allow `HEAD` on `/zd-movies-non-english/movies-non-english*`

then you could do something like this:

Create another key with that same prefix `/zd-movies-non-english/movies-non-englishabcd`

and now you're have access to both `/zd-movies-non-english/movies-non-english` and `/zd-movies-non-english/movies-non-englishabcd
`

**Comment by ncw** [comment]
> I think that if you relax the rules, technically you can open up a vector for exploring data that doesn't belong to that user.

Absolutely. I just wanted to confirm that that was the problem

The result of `git bisect` would be most useful :-)

**Comment by darthShadow** [comment]
> The result of git bisect would be most useful :-)

Will get it done sometime tomorrow.

**Comment by darthShadow** [comment]
@ncw This is the bad commit: 53aa03cc44b2a4250f0a1d162e10ffea5b38db9b

**Comment by ncw** [comment]
Give this a try

[v1.55.0-beta.5163.c781c26b0.fix-4990-s3-dir](https://beta.rclone.org/branch/fix-4990-s3-dir/v1.55.0-beta.5163.c781c26b0.fix-4990-s3-dir/) on branch [fix-4990-s3-dir](https://github.com/rclone/rclone/tree/fix-4990-s3-dir) (uploaded in 15-30 mins)

What it does is not check to see if the path passed in is an object if it ends with `/`.

This should fix your problem (provided you keep passing the path in with a trailing `/`) and it will also make s3 more efficient in the general case (one less transaction) when users pass in a `/`.

I should probably do this to **all** the backends, but let's see if it fixes your problem first!

**Comment by darthShadow** [comment]
Great, that works for paths ending with a slash and mostly solves my issue.

However, it's still broken for paths not ending with a slash which work perfectly fine in 1.53.4.

Not sure if anything more can be done there?

**Comment by ncw** [comment]
> Great, that works for paths ending with a slash and mostly solves my issue.

Great

> However, it's still broken for paths not ending with a slash which work perfectly fine in 1.53.4.
> 
> Not sure if anything more can be done there?

The problem is that rclone doesn't know whether the path you gave it is a file or directory.

The old code did a HEAD and if no errors were returned then assumed it was a file.

The new code checks the errors more carefully.

The problem error is permission denied. That could mean the object exists but you can't read it or you can't read the path and I'm not telling you about the object.

Previously we ignored it, now we report it.

Maybe we should go back to the old logic, no error implies object, anything else a directory.

What do you think?

Also can you set these filters up with anything other than minio?

**Comment by darthShadow** [comment]
> What do you think?

Perhaps revert it and put a deprecation notice and corresponding log to indicate an upcoming change in 1.55?

> Also can you set these filters up with anything other than minio?

Should be possible but I am not familiar with the exact steps. I can get you the exact policy if that helps?

**Comment by ncw** [comment]
OK I've reverted to assuming any errors mean the object isn't a directory

[v1.55.0-beta.5165.deed23e8b.fix-4990-s3-dir](https://beta.rclone.org/branch/fix-4990-s3-dir/v1.55.0-beta.5165.deed23e8b.fix-4990-s3-dir/) on branch [fix-4990-s3-dir](https://github.com/rclone/rclone/tree/fix-4990-s3-dir) (uploaded in 15-30 mins)

This should be backwards compatible behaviour.

Can you give it a test?

This is a candidate for a point release I think.

**Comment by darthShadow** [comment]
Works perfectly, thanks.

Do you want to add a deprecation notice for this behaviour too?

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54.1

I don't think we need to call out the change in behaviour as we are reverting it back to what it was before.

In fact I made the same change to b2 recently too!

**Comment by brvi** [comment]
Thank you very much @ncw and @darthShadow!
